libsolidity/codegen/CompilerContext.cpp

```diff
@@ -130,14 +130,27 @@ void CompilerContext::addVariable(VariableDeclaration const& _declaration,
 	m_localVariables[&_declaration].push_back(unsigned(m_asm->deposit()) - _offsetToCurrent);
 }
 
-void CompilerContext::removeVariable(VariableDeclaration const& _declaration)
+void CompilerContext::removeVariable(Declaration const& _declaration)
 {
 	solAssert(m_localVariables.count(&_declaration) && !m_localVariables[&_declaration].empty(), "");
 	m_localVariables[&_declaration].pop_back();
 	if (m_localVariables[&_declaration].empty())
 		m_localVariables.erase(&_declaration);
 }
 
+void CompilerContext::removeVariablesAboveStackHeight(unsigned _stackHeight)
+{
+	vector<Declaration const*> toRemove;
+	for (auto _var: m_localVariables)
+	{
+		solAssert(!_var.second.empty(), "");
+		if (_var.second.back() >= _stackHeight)
```

Comment:
Might it make sense to add two safety measures?
 - assert here that `_var.second.back()` is at most the current stack height
 - assert that `m_localVariables` is empty at the end of the function

Reply:
`m_localVariables` doesn't necessarily have to be empty in the end of the function, since it's called every time a scope closes

Reply:
Or did you mean in the end of visit to `FunctionDefinition` in `ContractCompiler`?

Reply:
There's a test breaking the first assertion suggestion, debugging it.

Reply:
Fixed